clippy_lints/src/floating_point_arithmetic.rs

```diff
@@ -359,6 +371,135 @@ fn check_mul_add(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
+            BinOpKind::Lt | BinOpKind::Le => is_zero(left) && are_exprs_equal(cx, right, test),
+            _ => false,
+        }
+    } else {
+        false
+    }
+}
+
+fn is_testing_negative(cx: &LateContext<'_, '_>, expr: &Expr<'_>, test: &Expr<'_>) -> bool {
+    if let ExprKind::Binary(Spanned { node: op, .. }, left, right) = expr.kind {
+        match op {
+            BinOpKind::Gt | BinOpKind::Ge => is_zero(left) && are_exprs_equal(cx, right, test),
+            BinOpKind::Lt | BinOpKind::Le => is_zero(right) && are_exprs_equal(cx, left, test),
+            _ => false,
+        }
+    } else {
+        false
+    }
+}
+
+fn are_exprs_equal(cx: &LateContext<'_, '_>, expr1: &Expr<'_>, expr2: &Expr<'_>) -> bool {
+    SpanlessEq::new(cx).ignore_fn().eq_expr(expr1, expr2)
+}
+
+/// Returns true iff expr is some zero literal
+fn is_zero(expr: &Expr<'_>) -> bool {
+    if let ExprKind::Lit(Lit { node: lit, .. }) = &expr.kind {
+        match lit {
+            LitKind::Int(0, _) => true,
+            LitKind::Float(symb, LitFloatType::Unsuffixed)
+            | LitKind::Float(symb, LitFloatType::Suffixed(FloatTy::F64)) => {
+                symb.as_str().parse::<f64>().unwrap() == 0.0
+            },
+            LitKind::Float(symb, LitFloatType::Suffixed(FloatTy::F32)) => symb.as_str().parse::<f32>().unwrap() == 0.0,
+            _ => false,
+        }
+    } else {
+        false
+    }
+}
+
+/// If the expressions are not opposites, return None
+/// Otherwise, return true if expr2 = -expr1, false if expr1 = -expr2 and return the positive
+/// expression
+fn are_opposites<'a>(
+    cx: &LateContext<'_, '_>,
+    expr1: &'a Expr<'a>,
+    expr2: &'a Expr<'a>,
+) -> Option<(bool, &'a Expr<'a>)> {
+    if let ExprKind::Block(
+        Block {
+            stmts: [],
+            expr: Some(expr1_inner),
+            ..
+        },
+        _,
+    ) = &expr1.kind
+    {
+        if let ExprKind::Block(
+            Block {
+                stmts: [],
+                expr: Some(expr2_inner),
+                ..
+            },
+            _,
+        ) = &expr2.kind
+        {
+            if let ExprKind::Unary(UnOp::UnNeg, expr1_neg) = &expr1_inner.kind {
+                if are_exprs_equal(cx, expr1_neg, expr2_inner) {
+                    return Some((false, expr2_inner));
+                }
+            }
+            if let ExprKind::Unary(UnOp::UnNeg, expr2_neg) = &expr2_inner.kind {
+                if are_exprs_equal(cx, expr1_inner, expr2_neg) {
+                    return Some((true, expr1_inner));
+                }
+            }
+        }
+    }
+    None
+}
+
+fn check_custom_abs(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
+    if let Some((cond, body, Some(else_body))) = higher::if_block(&expr) {
+        if let Some((expr1_pos, body)) = are_opposites(cx, body, else_body) {
+            let pos_abs_sugg = (
+                "This looks like you've implemented your own absolute value function",
```

Comment:
```suggestion
                "manual implementation of the `abs` method",
```

Don't address the programmer directly with `you`/`your`/... Also error, lint, and help messages start lowercase and don't have punctuation at the end.